### PR TITLE
Add QBF example programs using DepQBF

### PR DESCRIPTION
**Pull request by aswiridoff**
I added two example programs (Coloring.hs, Synthesis.hs) to examples/qbf. The first one solves a graph coloring problem, and the second one solves a circuit synthesis problem.

### CODE REVIEW

File: examples/qbf/Coloring.hs [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    y <- xs
+    return $ r!(x,y)
+
+-- | select 2 [1..4] = [[1,2],[1,3],[1,4],[2,3],[2,4],[3,4]]
+select :: Int -> [a] -> [[a]]
+select 0 _ = [[]]
+select _ [] = []
+select k (x:xs) = map (x:) (select (k-1) xs) <> select k xs
+
+-- | Given a relation r with domain a und codomain b, check if r matches every element in a
```
Comment (RyanGlScott) [inline examples/qbf/Coloring.hs:91]:
```suggestion
-- | Given a relation r with domain a and codomain b, check if r matches every element in a
```

File: examples/qbf/Coloring.hs [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+  let n = 22; c = 4; k = 3
+  formulaSize $ problem n c k
+  solve $ problem n c k
+
+-- | not 4-colorable and K4-free (11 nodes)
+variant_b :: IO ()
+variant_b = do
+  let n = 11; c = 4; k = 4
+  formulaSize $ problem n c k
+  solve $ problem n c k
```
Comment (RyanGlScott) [inline examples/qbf/Coloring.hs:42]: Neither of these `variant`s are actually used at runtime. Should they be?

File: examples/qbf/Coloring.hs [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+  let n = 11 -- number of nodes
+      c = 3  -- not 3-colorable
+      k = 3  -- K3-free
+  formulaSize $ problem n c k
+  solve $ problem n c k
+
+-- | not 4-colorable and K3-free (21 or 22 nodes?)
+variant_a :: IO ()
+variant_a = do
+  let n = 22; c = 4; k = 3
```
Comment (RyanGlScott) [inline examples/qbf/Coloring.hs:33]: Style nit pick:

```suggestion
  let n = 22
      c = 4
      k = 3
```

This is also more consistent with `main` above.

File: examples/qbf/Coloring.hs [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+  formulaSize $ problem n c k
+  solve $ problem n c k
+
+-- | not 4-colorable and K4-free (11 nodes)
+variant_b :: IO ()
+variant_b = do
+  let n = 11; c = 4; k = 4
```
Comment (RyanGlScott) [inline examples/qbf/Coloring.hs:40]: Similarly:

```suggestion
  let n = 11
      c = 4
      k = 4
```

File: examples/qbf/Coloring.hs [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+  formulaSize $ problem n c k
+  solve $ problem n c k
+
+
+solve :: StateT QSAT IO (Relation Int Int) -> IO ()
+solve p = do
+  result <- solveWith (depqbfPathArgs "depqbf" ["--qdo", "--no-dynamic-nenofex"]) $ p
+  case result of
+    (Satisfied, Just r) -> mapM_ putStrLn $ [table r, show $ edgesA r]
```
Comment (RyanGlScott) [inline examples/qbf/Coloring.hs:49]: These uses of `$` are redundant:

```suggestion
  result <- solveWith (depqbfPathArgs "depqbf" ["--qdo", "--no-dynamic-nenofex"]) p
  case result of
    (Satisfied, Just r) -> mapM_ putStrLn [table r, show $ edgesA r]
```

File: examples/qbf/Coloring.hs [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+  => ((a,b),(a,b)) -> m (Relation a b)
+universally_quantified_relation bnd = do
+  pairs <- sequence $ do
+    p <- A.range bnd
+    return $ do
+      x <- forall_
+      return (p,x)
+  return $ build bnd pairs
+
+-- | @has_k n r@ encodes the constraint, that @r@ has a complete subgraph with @n@ nodes.
```
Comment (RyanGlScott) [inline examples/qbf/Coloring.hs:76]:
```suggestion
-- | @has_k n r@ encodes the constraint that @r@ has a complete subgraph with @n@ nodes.
```

File: examples/qbf/Coloring.hs [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+-- | Given a relation r with domain a und codomain b, check if r matches every element in a
+-- to exactly one element in b.
+is_coloring :: Ix a => Relation a a -> Bit
+is_coloring c = and $ do
+  i <- domain c
+  return $ exactly 1 $ do
+    j <- codomain c
+    return $ c!(i,j)
+
+-- | @proper c r@ encodes the constraint, that @c@ is a proper coloring for @r@.
```
Comment (RyanGlScott) [inline examples/qbf/Coloring.hs:100]:
```suggestion
-- | @proper c r@ encodes the constraint that @c@ is a proper coloring for @r@.
```

File: examples/qbf/Coloring.hs [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+  (p,q) <- indices r
+  j <- codomain col
+  return $ r!(p,q) ==> not $ col!(p,j) && col!(q,j)
+
+
+edgesA :: (Ix a, Ix b) => Array (a,b) Bool -> [(a,b)]
+edgesA a = [ i | (i, b) <- A.assocs a, b == True]
+
+formulaSize :: StateT QSAT Identity a -> IO ()
+formulaSize p = mapM_ C.putStrLn $ take 2 $ B.split 10 $ qdimacsQSAT $ p
```
Comment (RyanGlScott) [inline examples/qbf/Coloring.hs:112]: Redundant `$` use:

```suggestion
formulaSize p = mapM_ C.putStrLn $ take 2 $ B.split 10 $ qdimacsQSAT p
```

File: examples/qbf/Synthesis.hs [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+-- | AIG where @exactly 2 [1,2,3]@ is equivalent to @8@
+aig_exactly :: IO ()
+aig_exactly = do
+  let n = 3 -- number of inputs
+      l = 5 -- number of steps (gates)
+      m = 1 -- number of outputs
+      f :: [Bit] -> [Bit]
+      f xs = [exactly 2 xs]
+  formulaSize $ problem f n l m
+  solve $ problem f n l m
```
Comment (RyanGlScott) [inline examples/qbf/Synthesis.hs:58]: Neither `aig_atmost` nor `aig_exactly` are used at runtime. Should they be?

File: examples/qbf/Synthesis.hs [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+
+
+solve :: StateT QSAT IO (Relation Int Int, Relation Int Int) -> IO ()
+solve p = do
+  result <- solveWith (depqbfPathArgs "depqbf" ["--qdo", "--no-dynamic-nenofex"]) $ p
```
Comment (RyanGlScott) [inline examples/qbf/Synthesis.hs:63]: Redundant `$` use:

```suggestion
  result <- solveWith (depqbfPathArgs "depqbf" ["--qdo", "--no-dynamic-nenofex"]) p
```

File: examples/qbf/Synthesis.hs [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+    _                       -> putStrLn "unsat"
+
+-- | @problem f n l m@ generates a QBF problem that encodes an AIG graph with @n@ inputs,
+-- @l@ gates and @m@ outputs which is equivalent to the boolean function @f@.
+problem :: Monad a => ([Bit] -> [Bit]) -> Int -> Int -> Int -> StateT QSAT a (Relation Int Int, Relation Int Int)
+problem f n l m = do
+  s <- relation ((n+1,1),(n+l,n+l)) -- selection variables: (x,y) is in s iff gate x takes gate/input y as an input
+  assert $ and $ do -- ensure that s is acyclic (DAG)
+    (i,j) <- indices s
+    guard $ (i <= j)
```
Comment (RyanGlScott) [inline examples/qbf/Synthesis.hs:75]: Redundant `$` use:

```suggestion
    guard (i <= j)
```

File: examples/qbf/Synthesis.hs [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+      x <- forall_
+      return (p,x)
+  return $ build bnd pairs
+
+
+edgesA :: (Ix a, Ix b) => Array (a,b) Bool -> [(a,b)]
+edgesA a = [ i | (i, b) <- A.assocs a, b == True]
+
+formulaSize :: StateT QSAT Identity a -> IO ()
+formulaSize p = mapM_ C.putStrLn $ take 2 $ B.split 10 $ qdimacsQSAT $ p
```
Comment (RyanGlScott) [inline examples/qbf/Synthesis.hs:109]: Redundant `$` use:

```suggestion
formulaSize p = mapM_ C.putStrLn $ take 2 $ B.split 10 $ qdimacsQSAT p
```

File: examples/qbf/Synthesis.hs [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+  pairs <- sequence $ do
+    p <- A.range bnd
+    return $ do
+      x <- forall_
+      return (p,x)
+  return $ build bnd pairs
+
+
+edgesA :: (Ix a, Ix b) => Array (a,b) Bool -> [(a,b)]
+edgesA a = [ i | (i, b) <- A.assocs a, b == True]
```
Comment (RyanGlScott) [inline examples/qbf/Synthesis.hs:106]: A bit simpler:

```suggestion
edgesA a = [ i | (i, True) <- A.assocs a]
```

File: examples/qbf/Coloring.hs [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+  return $ and $ do
+    (x:xs) <- tails xss
+    y <- xs
+    return $ r!(x,y)
+
+-- | select 2 [1..4] = [[1,2],[1,3],[1,4],[2,3],[2,4],[3,4]]
+select :: Int -> [a] -> [[a]]
+select 0 _ = [[]]
+select _ [] = []
+select k (x:xs) = map (x:) (select (k-1) xs) <> select k xs
```
Comment (RyanGlScott) [inline examples/qbf/Coloring.hs:89]: You can fix the pre-GHC 8.4 CI errors by making this change:

```suggestion
select k (x:xs) = map (x:) (select (k-1) xs) ++ select k xs
```

**Review by RyanGlScott — COMMENTED** [review]
Thanks, @aswiridoff! I have mostly minor suggestions.

**Comment by RyanGlScott** [comment]
Yes, the need for a specific version of `depqbf` is unfortunate. I wonder if there is a way we can query the version number and pass different default flags to `depqbf` to make it work for both old and new versions. (This isn't something that we need to do in this PR, but it would be a nice improvement for later.)

File: examples/qbf/Coloring.hs (hunk 2 above)
Comment (aswiridoff) [reply examples/qbf/Coloring.hs]: I added a small parser instead.

File: examples/qbf/Synthesis.hs (hunk 9 above)
Comment (aswiridoff) [reply examples/qbf/Synthesis.hs]: I used one as the main function and removed the other, but mentioned the corresponding problem in the comments.

**Review by RyanGlScott — APPROVED** [review]
Thanks, @aswiridoff! I've opened https://github.com/ekmett/ersatz/issues/94 to track the issue in https://github.com/ekmett/ersatz/pull/93#issuecomment-2092767109, which can be fixed independently.

Let's land this!

**Pull request merged**